Document pipeline registry and automatic discovery (#1788)
kedro-org/kedro: docs/source/nodes_and_pipelines/pipeline_registry.md

> # The pipeline registry

Reviewer: ```suggestion
# Pipeline registry
```
(very minor, feel free to ignore; personally I prefer how you have it now but it's more consistent with the style of most of our headings)

Author: I referenced [The Data Catalog](https://kedro.readthedocs.io/en/0.18.2/data/data_catalog.html), but I see how it's not consistent. Maybe another opportunity for consistency across the board.

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that return a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Reviewer 2: ```suggestion
Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:
```

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Reviewer: Since we plan to update the spaceflights starter to use `find_packages`, maybe we should link this to the 0.18.2 branch rather than `main`.

Author: Ah, I get your point. I've created a 0.18.3 branch (which, I guess, would be replaced by a 0.18.3 tag).

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Author: ```suggestion
Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. This, in turn, populates the `pipelines` variable in [`kedro.framework.project`](/kedro.framework.project) that the Kedro CLI and plugins use to access project pipelines. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:
```

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Author: Done!

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Author: ```suggestion
Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/0.18.3/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:
```

```python
def register_pipelines() -> Dict[str, Pipeline]:
    """Register the project's pipelines.

    Returns:
        A mapping from pipeline names to ``Pipeline`` objects.
    """
    data_processing_pipeline = dp.create_pipeline()
    data_science_pipeline = ds.create_pipeline()

    return {
        "__default__": data_processing_pipeline + data_science_pipeline,
        "data_processing": data_processing_pipeline,
        "data_science": data_science_pipeline,
    }
```

```{note}
The order in which you add the pipelines together is not significant (`data_science_pipeline + data_processing_pipeline` would produce the same result), since Kedro automatically detects the data-centric execution order for all the nodes in the resulting pipeline.
```

## Pipeline autodiscovery

> In the above example, you need to update the `register_pipelines()` function whenever you create a pipeline that should be returned as part of the project's pipelines. Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]`:

Reviewer 2: Should this be a link: `find_pipelines()`[/kedro.framework.project.find_pipelines]` ? Or just formatted?

Author: Ah, oops, nice catch!
```suggestion
In the above example, you need to update the `register_pipelines()` function whenever you create a pipeline that should be returned as part of the project's pipelines. Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]:
```

> In the above example, you need to update the `register_pipelines()` function whenever you create a pipeline that should be returned as part of the project's pipelines. Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]:

Reviewer: Following the above comment, this would then link to the file in the `main` branch.

Author: @AntonyMilneQB Sorry, I don't think I follow what needs to change here.

Reviewer: So my reading of the documentation as you have it is as follows.

> For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines... [example code as it was in 0.18.2]

and then you say

> Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]: [example code as it is in 0.18.3]

So my thinking was the first of these examples should link to the pipeline_registry.py file as it is in 0.18.2. Otherwise when I look at kedro's docs for 0.18.3, the link you follow to the `register_pipelines` function (that currently goes to `main`) will point to code using `find_pipelines` that doesn't look like the example code directly below the text. The example code you have there is copied and pasted from the spaceflights 0.18.2 starter. Then I was also thinking that the "example code as it is in 0.18.3" should point to `main` so it's always linked to the latest version of the `register_pipelines` file.

Does that make sense? Ultimately I don't think the kedro versions really matter here; what matters is the idea that doing the manual pipeline definitions of the second example is equivalent to the automatic `find_pipelines` in the second example (which is clear I think). I just thought that you were trying to link to the files that had exactly those code snippets, and the one that points to `main` was actually meant to point to 0.18.2.

```python
def register_pipelines() -> Dict[str, Pipeline]:
    """Register the project's pipelines.

    Since Kedro 0.18.3, projects can use the ``find_pipelines`` function
    to autodiscover pipelines. However, projects that require more fine-

>     grained control can still construct the pipeline mapping without it.

Author: ```suggestion
    grained control can still construct the pipeline mapping without it.
    
    For detailed information about how pipeline autodiscovery works, see
    https://kedro.readthedocs.io/en/stable/nodes_and_pipelines/pipeline_registry.html?#pipeline-autodiscovery

```

Reviewer 3: I think a reference link would work, my main point is we should include this information within the `find_pipeline` function, maybe just a link that explains what are the expected structure.

This is the current docstring in `find_pipeline`
```
"""
Automatically find modular pipelines having a ``create_pipeline``
    function. By default, projects created using Kedro 0.18.3 and higher
    call this function to autoregister pipelines upon creation/addition.
    Returns:
        A generated mapping from pipeline names to ``Pipeline`` objects.
    Warns:
        UserWarning: When a module does not expose a ``create_pipeline``
            function, the ``create_pipeline`` function does not return a
            ``Pipeline`` object, or if the module import fails up front.
 """
```

Author: Added a link.

    Returns:
        A mapping from pipeline names to ``Pipeline`` objects.
    """
    pipelines = find_pipelines()
    pipelines["__default__"] = sum(pipelines.values())
    return pipelines
```

> Under the hood, the `find_pipelines()` function traverses the `src/<package_name>/pipelines/` directory, returning a mapping from pipeline directory name to [`Pipeline`](/kedro.pipeline.Pipeline) object by:

Reviewer 3: I think this would be useful to include in the `find_pipelines()` docstring as well, but maybe you want to do it in the other PR directly.

> Under the hood, the `find_pipelines()` function traverses the `src/<package_name>/pipelines/` directory, returning a mapping from pipeline directory name to [`Pipeline`](/kedro.pipeline.Pipeline) object by:

Reviewer: ```suggestion
Under the hood, the `find_pipelines()` function traverses the `src/<package_name>/pipelines/` directory and returns a mapping from pipeline directory name to [`Pipeline`](/kedro.pipeline.Pipeline) object by:
```

> 1. importing the `<package_name>.pipelines.<pipeline_name>` module

Author: Thanks! I Googled for general practices, but didn't realize it's called out in the detailed contribution guidelines. :)

> 1. importing the `<package_name>.pipelines.<pipeline_name>` module

Author: ```suggestion
1. Importing the `<package_name>.pipelines.<pipeline_name>` module
```

> 1. calling the `create_pipeline()` function exposed by the `<package_name>.pipelines.<pipeline_name>` module

Author: ```suggestion
2. Calling the `create_pipeline()` function exposed by the `<package_name>.pipelines.<pipeline_name>` module
```

> 1. calling the `create_pipeline()` function exposed by the `<package_name>.pipelines.<pipeline_name>` module
> 1. validating that the constructed object is a [`Pipeline`](/kedro.pipeline.Pipeline)

Reviewer 3: ```suggestion
2. calling the `create_pipeline()` function exposed by the `<package_name>.pipelines.<pipeline_name>` module
3. validating that the constructed object is a [`Pipeline`](/kedro.pipeline.Pipeline)
```

Author: Thanks! I should have checked what was done across the codebase.

> 1. validating that the constructed object is a [`Pipeline`](/kedro.pipeline.Pipeline)

Author: ```suggestion
3. Validating that the constructed object is a [`Pipeline`](/kedro.pipeline.Pipeline)
```

> If any of these steps fail, `find_pipelines()` raises an appropriate warning and skips the current pipeline but continues traversal.

Reviewer 2: Maybe add an example of what this warning looks like?

Author: The warning depends on which of the above steps fails (or, in the case of "calling the `create_pipeline()` function exposed by the `<package_name>.pipelines.<pipeline_name>` module", which part of the step fails), so it's quite a lot of warnings. I think it would clutter the page to add all of them, and I think the warning messages are fairly self-explanatory if they occur. WDYT?

Reviewer 2: Okay that make sense! I thought maybe all of them include the "Nothing defined therein will be returned by find_pipelines" part, so they know what they look out for. But I'm fine leaving this as is.

> The mapping returned by `find_pipelines()` can be modified. For example, to add a data engineering pipeline that isn't part of the default pipeline, add it to the dictionary *after* constructing the default pipeline:

Reviewer 4: Just for the record, this is something very common for me when I use namespaces or modular pipelines: I duplicate and modify slightly the pipeline at this moment. 

Combining pipelines as described below is quite common too, sometimes I want to be able to run two pipelines separately (e.g. de, ds) or at the same time (de+ds).

> The mapping returned by `find_pipelines()` can be modified. For example, to add a data engineering pipeline that isn't part of the default pipeline, add it to the dictionary *after* constructing the default pipeline:

Reviewer: I think it would be nice to have another example here, or maybe just a note to suggest some other things users might like to do:
* filter pipelines like `pipelines["data_processing"] = pipelines["data_processing"].filter(tags=..., from_nodes=...)`
* put this before the definition of `__default__` if you want it reflected when you do `kedro run` without specifying a pipeline

Also I think we should have a sentence that explains what that `pipelines["__default__"]` does. Maybe after you say "The mapping returned by `find_pipelines()` can be modified.", something like "The default template also defines a `__default__` pipeline (executed by `kedro run` when blah blah). This is the concatenation of all blah blah blah".

> The mapping returned by `find_pipelines()` can be modified. For example, to add a data engineering pipeline that isn't part of the default pipeline, add it to the dictionary *after* constructing the default pipeline:

Author: How about https://github.com/kedro-org/kedro/commit/4db786b5d2316fb32e73df420d3b1fc006944970? @AntonyMilneQB I didn't find it necessary to add examples of `filter`, `from_nodes`, etc.; I think this is sufficiently documented elsewhere, and isn't exactly pertinent. I did go ahead and re-explain the default pipeline, providing a link (even though one could argue that need not be done here).

> The mapping returned by `find_pipelines()` can be modified. For example, to add a data engineering pipeline that isn't part of the default pipeline, add it to the dictionary *after* constructing the default pipeline:

Reviewer: @deepyaman This looks good. The main thing I thought might require additional explanation though is what `sum(pipelines.values())` does, since this is quite new kedro syntax and maybe not obvious to a beginner user. Something about how it concatenates together all the nodes from pipelines in the `pipelines` dictionary. 

But this is not a big thing, feel free to leave out if you think it's not useful.

# The pipeline registry

> Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. This, in turn, populates the `pipelines` variable in [`kedro.framework.project`](/kedro.framework.project) that the Kedro CLI and plugins use to access project pipelines. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the pipeline registry in the [Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/tree/main/spaceflights) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:

Author: @AntonyMilneQB Like so?

```suggestion
Projects generated using Kedro 0.17.2 or later define their pipelines in `src/<package_name>/pipeline_registry.py`. This, in turn, populates the `pipelines` variable in [`kedro.framework.project`](/kedro.framework.project) that the Kedro CLI and plugins use to access project pipelines. The `pipeline_registry` module must contain a top-level `register_pipelines()` function that returns a mapping from pipeline names to [`Pipeline`](/kedro.pipeline.Pipeline) objects. For example, the [pipeline registry in the Kedro starter for the completed spaceflights tutorial](https://github.com/kedro-org/kedro-starters/blob/0.18.2/spaceflights/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D/pipeline_registry.py) could define the following `register_pipelines()` function that exposes the data processing pipeline, the data science pipeline, and a third, default pipeline that combines both of the aforementioned pipelines:
```

Reviewer: That's what I was thinking 👍

```python
def register_pipelines() -> Dict[str, Pipeline]:
    """Register the project's pipelines.

    Returns:
        A mapping from pipeline names to ``Pipeline`` objects.
    """
    data_processing_pipeline = dp.create_pipeline()
    data_science_pipeline = ds.create_pipeline()

    return {
        "__default__": data_processing_pipeline + data_science_pipeline,
        "data_processing": data_processing_pipeline,
        "data_science": data_science_pipeline,
    }
```

As a reminder, [running `kedro run` without the `--pipeline` option runs the default pipeline](./run_a_pipeline.md#run-a-pipeline-by-name).

```{note}
The order in which you add the pipelines together is not significant (`data_science_pipeline + data_processing_pipeline` would produce the same result), since Kedro automatically detects the data-centric execution order for all the nodes in the resulting pipeline.
```

## Pipeline autodiscovery

> In the above example, you need to update the `register_pipelines()` function whenever you create a pipeline that should be returned as part of the project's pipelines. Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]:

Author: ```suggestion
In the above example, you need to update the `register_pipelines()` function whenever you create a pipeline that should be returned as part of the project's pipelines. Since Kedro 0.18.3, you can achieve the same result with less code using `find_pipelines()`[/kedro.framework.project.find_pipelines]. The [updated pipeline registry](https://github.com/kedro-org/kedro-starters/blob/main/spaceflights/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D/pipeline_registry.py) contains no project-specific code:
```

Reviewer: Yes indeed.
